File: 1.29/integrals.exs
defmodule Integral do
  def simple(f, a, b, dx) do
    add_dx = fn x -> x + dx end
    Sum.calc(f, a + dx / 2, add_dx, b) * dx
  end

  def simpson(f, a, b, n) do
    h = (b - a) / n
    add_h = fn x -> x + h end

    simpson_member = (fn(x) ->
      k = round((x - a) / h)

      coef = cond do
        k == 0 || k == n -> 1
        rem(k, 2) == 0   -> 2
        true             -> 4
      end
      
      # IO.puts "k=#{k}, coef=#{coef}, member=#{coef * f.(x)}"
      coef * f.(x)
    end)

    Sum.calc(simpson_member, a, add_h, b) * h / 3
  end
end

defmodule Sum do
  def calc(term, a, next, b) do
    if (a - b) > (next.(a) - a) * 0.3 do
      0
    else
      term.(a) + calc(term, next.(a), next, b)
    end
  end
end

File: 1.29/integrals.rs
pub fn simple(a: f64, b: f64, f: &dyn Fn(f64) -> f64, dx: f64) -> f64 {
    let add_dx = |x: f64| -> f64 { x + dx };
    sum(a + dx / 2.0, b, f, &add_dx) * dx
}

pub fn simpson(a: f64, b: f64, f: &dyn Fn(f64) -> f64, n: i32) -> f64 {
    let h = (b - a) / (n as f64);
    let add_h = |x: f64| -> f64 { x + h };

    let simpson_member = |x: f64| -> f64 {
        let k: i32 = ((x - a) / h).round() as i32;
        let coef: f64;

        if k == 0 || k == n {
            coef = 1.0;
        } else if k % 2 == 0 {
            coef = 2.0;
        } else {
            coef = 4.0;
        }

        coef * f(x)
    };

    sum(a, b, &simpson_member, &add_h) * h / 3.0
}

fn sum(a: f64, b: f64, term: &dyn Fn(f64) -> f64, nexta: &dyn Fn(f64) -> f64) -> f64 {
    if a - b > (nexta(a) - a) * 0.3 {
        0f64
    } else {
        term(a) + sum(nexta(a), b, term, nexta)
    }
}


